### ISSUE TITLE
Get Emiggen and Ray working on all available compute clusters

### ISSUE THREAD

**Issue by justaddcoffee** [monarch-initiative/embiggen#173]
We have access to at least these clusters:

Google Cloud - in progress - Fengchen working on this
Lawrencium - in progress - Marcin working on this
NERSC/Cori - in progress - Justin working on this
JAX cluster - working

Ideally it'd be great to get tooling working to run Emiggen and Ray on any/all of these clusters to find optimal hyperparameters. 

Luca's code to deploy ray on clusters is here:
https://github.com/LucaCappelletti94/distributed_bayesian_optimization

We had a hackathon on May 22 where we worked a bit, and made progress on this goal.

**Comment by justaddcoffee** [comment]
Getting close to working on NERSC/Cori, but I'm seeing this error: 

```
(venv) (base) jreese@cori03:/global/cscratch1/sd/jreese/distributed_bayesian_optimization> cat slurm-30987680.out
srun: fatal: Can not execute ray
srun: fatal: Can not execute ray
srun: fatal: Can not execute ray
srun: fatal: Can not execute ray
srun: fatal: Can not execute ray
Traceback (most recent call last):
  File "bayesian_opt.py", line 5, in <module>
    from distributed_bayesian_optimization import distributed_bo
  File "/global/cscratch1/sd/jreese/distributed_bayesian_optimization/distributed_bayesian_optimization/__init__.py", line 1, in <module>
    from .distributed_bo import distributed_bo
  File "/global/cscratch1/sd/jreese/distributed_bayesian_optimization/distributed_bayesian_optimization/distributed_bo.py", line 1, in <module>
    from ray import tune
ImportError: cannot import name 'tune' from 'ray' (unknown location)
```

@LucaCappelletti94 ideas?

**Comment by richardliaw** [comment]
why is there an unknown location, are the right modules loaded? maybe try something simple like `import ray; ray.init()`?

**Comment by LucaCappelletti94** [comment]
In the meantime @justaddcoffee, I have updated the [setup and test guide](https://github.com/LucaCappelletti94/distributed_bayesian_optimization). If you have 15 minutes to spare, we could have a small test run.

**Comment by justaddcoffee** [comment]
@LucaCappelletti94 sure, let's sync up, maybe early next week

**Comment by justaddcoffee** [comment]
> why is there an unknown location, are the right modules loaded? maybe try something simple like `import ray; ray.init()`?

Thanks @richardliaw

**Comment by fengchenLBL** [comment]
Successfully installed `distributed_bayesian_optimization` on a [Google Cloud cluster](https://www.qwiklabs.com/focuses/10390?parent=catalog). @justaddcoffee, I shared a google doc with you.   

Install:
```
# login to the gcp master node
gcloud compute ssh g1-login0 --zone=<ZONE>

git clone https://github.com/LucaCappelletti94/distributed_bayesian_optimization.git
cd distributed_bayesian_optimization

sudo yum install python-devel python3-devel gcc-c++

python3 -m venv my_venv
source my_venv/bin/activate
bash setup.sh
```
Run a test:
```
# In bayesian_test.sh,  add "source my_venv/bin/activate"
sbatch bayesian_test.sh
```
Got the results in __BO.csv__. 
However, there are some errors in the slurm log file.
@LucaCappelletti94 , Did I miss something? Thanks.
```
2020-05-28 00:27:39,603 WARNING scripts.py:318 -- The --redis-port argument will be deprecated soon. Please use --port instead.
2020-05-28 00:27:39,603 INFO scripts.py:393 -- Using IP address 10.0.0.9 for this node.
2020-05-28 00:27:39,607 INFO resource_spec.py:212 -- Starting Ray with 4.05 GiB memory available for workers and up to 2.05 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
2020-05-28 00:27:39,850 WARNING services.py:924 -- Redis failed to start, retrying now.
2020-05-28 00:27:40,690 INFO services.py:1166 -- View the Ray dashboard at localhost:8265
2020-05-28 00:27:40,803 INFO scripts.py:423 --
Started Ray on this node. You can add additional nodes to the cluster by calling

    ray start --address='10.0.0.9:6379' --redis-password='41e207f1-a2ba-4385-b67f-6d475880e8cd'

from the node you wish to add. You can connect a driver to the cluster from Python by running

    import ray
    ray.init(address='auto', redis_password='41e207f1-a2ba-4385-b67f-6d475880e8cd')

If you have trouble connecting from a different machine, check that your firewall is configured properly. If you wish to terminate the processes that have been started, run

    ray stop
2020-05-28 00:28:28,434 ERROR trial_runner.py:519 -- Trial custom_loss_185d4b1a: Error processing event.
Traceback (most recent call last):
```

**Comment by LucaCappelletti94** [comment]
Do the BO results show a convergence? I mean, do they get very small by the end?

I'll look into the errors and I'll try to understand what is happening there.

**Comment by justaddcoffee** [comment]
Thanks @fengchenLBL! this is great

@LucaCappelletti94, maybe we 3 can chat when you get a chance about that error?

**Comment by LucaCappelletti94** [comment]
Sure, where would you prefer to chat?

**Comment by fengchenLBL** [comment]
@LucaCappelletti94 , this is the BO.csv I got from the test:
```
,my_custom_loss,time_this_iter_s,done,timesteps_total,episodes_total,training_iteration,experiment_id,date,timestamp,time_total_s,pid,hostname,node_ip,time_since_restore,timesteps_since_restore,iterations_since_restore,trial_id,experiment_tag,config/x,logdir
0,56.112120250478576,0.0022580623626708984,False,,,1,d1fcc2935df845f9a59d28b644bc0474,2020-05-28_00-27-59,1590625679,0.0022580623626708984,1631,g1-compute-0-0,10.0.0.9,0.0022580623626708984,0,1,14e494fc,1_x=7.4908,7.49080237694725,/home/user/distributed_bayesian_optimization/gaussian_process/gaussian_process/custom_loss_1_x=7.4908_2020-05-28_00-27-58_i5xdrty
1,361.5430769649951,0.0003578662872314453,True,,,1,fda19dbfa66c4c8884b1700cd5a1816f,2020-05-28_00-27-59,1590625679,0.0003578662872314453,1630,g1-compute-0-0,10.0.0.9,0.0003578662872314453,0,1,14e4e0d8,2_x=19.014,19.01428612819832,/home/user/distributed_bayesian_optimization/gaussian_process/gaussian_process/custom_loss_2_x=19.014_2020-05-28_00-27-585abegzqd
2,214.32605233943943,0.006490468978881836,True,,,1,dd6276fe20fc4847977dbf044d398c5e,2020-05-28_00-28-02,1590625682,0.006490468978881836,1729,g1-compute-0-0,10.0.0.9,0.006490468978881836,0,1,14e5109e,3_x=14.64,14.639878836228101,/home/user/distributed_bayesian_optimization/gaussian_process/gaussian_process/custom_loss_3_x=14.64_2020-05-28_00-27-599g14_kwm
3,143.35679228043742,0.02000880241394043,True,,,1,dfd7fa6e73bc43808d4d2142866eca0a,2020-05-28_00-28-02,1590625682,0.02000880241394043,1730,g1-compute-0-0,10.0.0.9,0.02000880241394043,0,1,14e54d66,4_x=11.973,11.973169683940732,/home/user/distributed_bayesian_optimization/gaussian_process/gaussian_process/custom_loss_4_x=11.973_2020-05-28_00-27-59ttjhkpdx
4,9.736726466202516,0.0007038116455078125,False,,,1,8a00c49f255a405792f5a2515c32fff5,2020-05-28_00-28-04,1590625684,0.0007038116455078125,1744,g1-compute-0-0,10.0.0.9,0.0007038116455078125,0,1,14e57912,5_x=3.1204,3.1203728088487304,/home/user/distributed_bayesian_optimization/gaussian_process/gaussian_process/custom_loss_5_x=3.1204_2020-05-28_00-28-02nr73mgv8
5,0.0,0.012756109237670898,False,,,1,9cf7be17e7b943f4b6ae567eb832d9a2,2020-05-28_00-28-26,1590625706,0.012756109237670898,1797,g1-compute-0-0,10.0.0.9,0.012756109237670898,0,1,184795f4,6_x=0.0,0.0,/home/user/distributed_bayesian_optimization/gaussian_process/gaussian_process/custom_loss_6_x=0.0_2020-05-28_00-28-25u3fu8kd1
6,0.0,0.013822317123413086,False,,,1,18b17a589ab245a2be7f2bd6174a727d,2020-05-28_00-28-29,1590625709,0.013822317123413086,1960,g1-compute-0-0,10.0.0.9,0.013822317123413086,0,1,18538a44,7_x=0.0,0.0,/home/user/distributed_bayesian_optimization/gaussian_process/gaussian_process/custom_loss_7_x=0.0_2020-05-28_00-28-2590r39qep
7,0.0,0.023410797119140625,False,,,1,892f37dffac649d195b95d44cc789fdb,2020-05-28_00-28-28,1590625708,0.023410797119140625,1958,g1-compute-0-0,10.0.0.9,0.023410797119140625,0,1,185d4b1a,8_x=0.0,0.0,/home/user/distributed_bayesian_optimization/gaussian_process/gaussian_process/custom_loss_8_x=0.0_2020-05-28_00-28-26bnfbav2f
8,0.0,0.005444526672363281,False,,,1,291d0a051279450ba768c853abd60e07,2020-05-28_00-28-30,1590625710,0.005444526672363281,1971,g1-compute-0-0,10.0.0.9,0.005444526672363281,0,1,1866228a,9_x=0.0,0.0,/home/user/distributed_bayesian_optimization/gaussian_process/gaussian_process/custom_loss_9_x=0.0_2020-05-28_00-28-28232zqj68
9,0.0,0.005241870880126953,False,,,1,acb6cb53dbd44cc99d7bf6e7cd78e551,2020-05-28_00-28-32,1590625712,0.005241870880126953,1994,g1-compute-0-0,10.0.0.9,0.005241870880126953,0,1,186bf214,10_x=0.0,0.0,/home/user/distributed_bayesian_optimization/gaussian_process/gaussian_process/custom_loss_10_x=0.0_2020-05-28_00-28-29dmy8jzgo
10,0.0,0.013014554977416992,False,,,1,03ea22ec0c2c4074839c7a4ef6ced259,2020-05-28_00-28-33,1590625713,0.013014554977416992,2011,g1-compute-0-0,10.0.0.9,0.013014554977416992,0,1,1876a164,11_x=0.0,0.0,/home/user/distributed_bayesian_optimization/gaussian_process/gaussian_process/custom_loss_11_x=0.0_2020-05-28_00-28-303gl_ybqv
11,0.0,0.0014150142669677734,False,,,1,ba3275700dee401991ee19e7acc59826,2020-05-28_00-28-34,1590625714,0.0014150142669677734,2025,g1-compute-0-0,10.0.0.9,0.0014150142669677734,0,1,187efa4e,12_x=0.0,0.0,/home/user/distributed_bayesian_optimization/gaussian_process/gaussian_process/custom_loss_12_x=0.0_2020-05-28_00-28-32gckjbihx
```

**Comment by justaddcoffee** [comment]
@fengchenLBL - thanks so much for the effort, this is great. 

> Sure, where would you prefer to chat?

hi @LucaCappelletti94 - could chat today at some point, or we can cat asynchronously on this ticket. In the meantime, does the output of Fengchen look reasonable? It looks to me like it's working, and the loss function steadily decreases...

**Comment by LucaCappelletti94** [comment]
@justaddcoffee yes that looks like it is working as expected! We still have the points collision, but those are to be addressed in Ray. I'm already working on a new pull request to patch that aspect with @richardliaw.

As for the chat, today I've really little time without meetings, would Sunday or Monday morning (your time) work for you?

**Comment by fengchenLBL** [comment]
@justaddcoffee @LucaCappelletti94  Good to know it's working as expected. Both Sunday and Monday work for me.

**Comment by richardliaw** [comment]
hey all, is this blocked on anything? anything you need from the ray side?

**Comment by LucaCappelletti94** [comment]
> hey all, is this blocked on anything? anything you need from the ray side?

Hi @richardliaw, currently we are still finishing up the codebase. We will start again experimenting with BO in the next few weeks and we'll let you know how it goes!

**Comment by LucaCappelletti94** [comment]
Closing this issue as stale, no Ray experimentation was executed in the end.
